src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -88,10 +89,17 @@ internal sealed class PeWriter
             // If PDB writer is given, we have to have PDB path.
             Debug.Assert(nativePdbWriterOpt == null || pdbPathOpt != null);
 
-            var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
-            var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
+            try
+            {
+                var peWriter = new PeWriter(context.Module.Properties, context.Module.Win32Resources, context.Module.Win32ResourceSection, pdbPathOpt, deterministic);
+                var mdWriter = FullMetadataWriter.Create(context, messageProvider, allowMissingMethodBodies, deterministic, getPortablePdbStreamOpt != null, cancellationToken);
 
-            return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+                return peWriter.WritePeToStream(mdWriter, getPeStream, getPortablePdbStreamOpt, nativePdbWriterOpt);
+            }
+            catch (Exception ex) when (!(ex is PdbWritingException || ex is ResourceException || ex is PermissionSetFileReadException))
```

Comment:
> Only errors that are converted from HResults by the marshaller.

Which can include I/O errors.  Your proposal would cause I/O errors on PDB to be diagnostics but I/O errors on PE to be exceptions.  This isn't consistent.  

> My problem with this change is that this catch covers the entire implementation of PE writer, metadata writer, PDB writer and user code

The code today is responsible for the number one source of Watson's precisely because it's not guarded.  Moving the guard up a level in the compiler both creates:

- Creates an inconsistency between PE and PDB writing errors
- Makes the hosting story more difficult to understand for consumers

The PR makes this simple: exceptions during writing will become diagnostics. 

> if we convert exceptions like ArgumentOutOfRange exception into a diagnostic that diagnostic will say "Error while emitting 'App': argument 'foo' is out of range". Huh?

This is the model .NET exceptions force you into.  It's already done in many places in the Roslyn code base today.  This is not a new problem.  When these are reported we file bugs and fix the issues.